crates/network-scanner/src/task_utils.rs

```diff
@@ -0,0 +1,108 @@
+use std::{net::IpAddr, sync::Arc, time::Duration};
+
+use dashmap::DashMap;
+
+use futures::Future;
+use tokio::sync::Mutex;
+
+use crate::{
+    ip_utils::{get_subnets, Subnet},
+    scanner::NetworkScanner,
+};
+
+type IpSender = tokio::sync::mpsc::Sender<(IpAddr, Option<String>)>;
+type IpReceiver = tokio::sync::mpsc::Receiver<(IpAddr, Option<String>)>;
+type PortSender = tokio::sync::mpsc::Sender<(IpAddr, Option<String>, u16)>;
+type PortReceiver = tokio::sync::mpsc::Receiver<(IpAddr, Option<String>, u16)>;
+#[derive(Debug, Clone)]
+pub(crate) struct TaskExecutionContext {
+    pub handles_sender: HandlesSender,
+    pub ip_sender: IpSender,
+    pub ip_receiver: Arc<Mutex<IpReceiver>>,
+
+    pub port_sender: PortSender,
+    pub port_receiver: Arc<Mutex<PortReceiver>>,
+
+    pub ip_cache: Arc<DashMap<IpAddr, Option<String>>>,
+    pub ports: Vec<u16>,
+
+    pub runtime: Arc<network_scanner_net::runtime::Socket2Runtime>,
+    pub ping_interval: Duration,     // in milliseconds
+    pub ping_timeout: Duration,      // in milliseconds
+    pub broadcast_timeout: Duration, // in milliseconds
+    pub port_scan_timeout: Duration, // in milliseconds
+    pub netbios_timeout: Duration,   // in milliseconds
+    pub subnets: Vec<Subnet>,
+}
+
+type HandlesReceiver = crossbeam::channel::Receiver<tokio::task::JoinHandle<anyhow::Result<()>>>;
+type HandlesSender = crossbeam::channel::Sender<tokio::task::JoinHandle<anyhow::Result<()>>>;
+#[derive(Debug)]
+pub(crate) struct TaskExecutionRunner {
+    pub(crate) context: TaskExecutionContext,
+    pub(crate) handles_receiver: HandlesReceiver,
+    pub(crate) handles_sender: HandlesSender,
+}
+
+impl TaskExecutionContext {
```

Comment:
This is created purely for the sake of readability, it was super hard of keep track of things without it.